# Search matched sum 리포트 과제

## 201811958 손휘준

### bubblesort 원리를 이용한 search_matched_sum_bubbleSort 함수구현 (bubblesort의 Time complexity : n^2)

In [2]:
def search_matched_sum_bubbleSort(arr : list, target : int):
    new_arr=[] # new_arr을 만든이유는 소팅하기전 기존배열의 인덱스정보도 함께 저장하기 위함입니다.
    for i,v in enumerate(arr):
        new_arr.append((i,v)) # new_arr에는 리스트내의 튜플형태로 (인덱스정보, 값)이 들어가게 됩니다.
    arr = bubbleSort(new_arr) 
    min_index=[] # 타겟에 대한 요소들의 인덱스를 모두 나타내기위해 리스트형태로 만들어 append를 사용할 예정입니다.
    for idx1, d1 in enumerate(arr): # search 함수자체의 time complexity는 이중 for문을 거치기때문에 n^2입니다.
        for idx2, d2 in enumerate(arr[idx1:], idx1): #enumerate 함수를 통해 idx1부터 start하여 그 후의 값을 봅니다.
            if d1[1]+d2[1] == target: # d1과 d2는 (sort하기전 배열에서 인덱스정보, 해당 값)을 가지고 있습니다.
                if d1[1]==d2[1]: # 예를 들어 target이 6인경우 3+3=6으로 3의 인덱스가 두번표시될수 있기때문에 서로 다른 두가지 요소들에 대해서는 표시하지않도록 처리해줍니다.
                    pass
                else:
                    min_index.append((d1[0],d2[0]))
    return "target={}".format(target), min_index

def bubbleSort(new_arr):  # 기존의 버블소트와 다르게 입력값인 new_arr은 리스트내에 튜플이 존재하는 형태입니다. 
    n = len(new_arr)
    for i in range(n): # bubblesort 함수만 본다면 time complexity는 이중 for문을 거치기때문에 n^2입니다.
        for j in range(0, n-i-1):
            if new_arr[j][1] > new_arr[j+1][1] : # new_arr 리스트 내의 튜플에 두번째 요소가 값을 나타내므로 [1]을 사용했습니다.
                new_arr[j], new_arr[j+1] = new_arr[j+1], new_arr[j]
    return new_arr


### search_matched_sum_bubbleSort 함수 테스트

In [3]:
arr = [10,9,8,7,6,5,4,3,2,1]
result = search_matched_sum_bubbleSort(arr,11) 
print(result) # 11을 타겟으로 한다면 값 (1,10), (2,9) (3,8), (4,7), (5,6)에 대한 인덱스쌍이 5개가 나올것입니다.

('target=11', [(9, 0), (8, 1), (7, 2), (6, 3), (5, 4)])


### quicksort 원리를 이용한 search_matched_sum_quickSort 함수구현 (quicksort의 Time complexity : nlog(n))

In [4]:
def search_matched_sum_quickSort(arr : list, target : int, low, high): # 위의 버블소트를 이용한 함수와 같은원리로 작성했습니다.
    new_arr=[]
    for i,v in enumerate(arr):
        new_arr.append((i,v))
    arr = quickSort(new_arr, low, high) 
    min_index=[]
    for idx1, d1 in enumerate(arr): # search 함수자체의 time complexity는 이중 for문을 거치기때문에 n^2입니다.
        for idx2, d2 in enumerate(arr[idx1:], idx1): 
            if d1[1]+d2[1] == target: 
                if d1[1]==d2[1]: 
                    pass
                else:
                    min_index.append((d1[0],d2[0]))
    return "target={}".format(target), min_index

def partition(new_arr, low, high):
    i = (low-1)        
    pivot = new_arr[high][1]  

    for j in range(low, high):
        if new_arr[j][1] <= pivot:

            i = i+1
            new_arr[i], new_arr[j] = new_arr[j], new_arr[i]

    new_arr[i+1], new_arr[high] = new_arr[high], new_arr[i+1]
    return (i+1)

def quickSort(new_arr, low, high):
    if len(new_arr) == 1: # quicksort 함수만 본다면 time complexity는 pivot을 기준으로 분할하며 진행하기때문에 nlog(n)입니다.
        return new_arr
    if low < high:
        pi = partition(new_arr, low, high)

        quickSort(new_arr, low, pi-1)
        quickSort(new_arr, pi+1, high)
    return new_arr


### search_matched_sum_quickSort 함수 테스트

In [5]:
arr = [10,9,8,7,6,5,4,3,2,1]
result = search_matched_sum_quickSort(arr,11,0,9) 
print(result) # 11을 타겟으로 한다면 값 (1,10), (2,9) (3,8), (4,7), (5,6)에 대한 인덱스쌍이 5개가 나올것입니다.

('target=11', [(9, 0), (8, 1), (7, 2), (6, 3), (5, 4)])


## ㅁ가설수립과 코드설명ㅁ

### search_matched_sum_bubbleSort, search_matched_sum_quickSort 두가지의 코드를 작성했습니다. 과제에서 요구하는 바는 배열에서 요소간의 덧셈을 통해 target과 같아지는 두 요소가 있다면 해당 요소들의 기존 배열의 인덱스를 쌍으로 도출하는 것이었습니다. 당연히 인덱스 쌍이 여러가지로 나올 수 있기 때문에 search를 하는 함수에서 이중 for문과 enurate함수를 통해 모든 쌍을 잡아낼 수 있도록 설계하였습니다. 그리고 기존의 배열의 인덱스 정보도 포함하여야 하기 때문에 기존의 인덱스정보와 값을 튜플형태로 저장하는 새로운 배열을 만들고 이를 sort하는 과정을 거쳤습니다. 두 search하는 함수 모두 이중 for문을 사용하였기 때문에 전체적인 time complexity는 N^2으로 볼 수 있지만 저는 search 하는 과정에서 n^2의 time complexity를 가지는 bubblesort와 nlog(n)의 time complexity를 가지는 quicksort를 통해 먼저 소팅을 하는 과정을 거쳐가기때문에 이 둘의 시간비교를 통해 n^2과 nlog(n)의 유의미한 차이를 얻을 수 있겠다고   생각했습니다.

### 따라서, 버블소트(n^2)를 이용한 search_matched_sum_bubbleSort이 퀵소트(nlog(n))를 이용한  search_matched_sum_quickSort보다 시간이 오래걸릴것이며, 배열의 개수가 증가한다면 그 시간차이가 더욱 많이날 것이라고 가설을 세웠습니다. 


### 1. 배열의 개수가 10, target이 10일때 시간을 측정

In [16]:
import random
import timeit
code='''
def test_func():
    n=10
    target=10
    random_arr = random.sample(range(1,n*2),n) # 1부터 n*2의 범위에서 무작위 숫자 n개를 추출하여 random_arr로 설정했습니다.
    result1 = search_matched_sum_bubbleSort(random_arr,target)
    return result1'''

def test_func():
    n=10
    target=10
    random_arr = random.sample(range(1,n*2),n)
    result1 = search_matched_sum_bubbleSort(random_arr,target)
    return result1

print(test_func())
print(timeit.timeit(stmt=code))
%timeit test_func()

('target=10', [(8, 4)])
0.17695209999999406
61.4 µs ± 5.87 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [17]:
import random
import timeit
code='''
def test_func():
    n=10
    target=10
    random_arr = random.sample(range(1,n*2),n)
    result2 = search_matched_sum_quickSort(random_arr, target, 0, n-1)
    return result2'''

def test_func():
    n=10
    target=10
    random_arr = random.sample(range(1,n*2),n)
    result2 = search_matched_sum_quickSort(random_arr, target, 0, n-1)
    return result2

print(test_func())
print(timeit.timeit(stmt=code))
%timeit test_func()

('target=10', [(3, 1)])
0.12228099999998676
55.8 µs ± 9.94 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


### 2. 배열의 개수가 1000, target이 10일때 시간을 측정

In [18]:
import random
import timeit
code='''
def test_func():
    n=1000
    target=10
    random_arr = random.sample(range(1,n*2),n)
    result1 = search_matched_sum_bubbleSort(random_arr,target)
    return result1'''

def test_func():
    n=1000
    target=10
    random_arr = random.sample(range(1,n*2),n)
    result1 = search_matched_sum_bubbleSort(random_arr,target)
    return result1

print(test_func())
print(timeit.timeit(stmt=code))
%timeit test_func()

('target=10', [])
0.09844480000003841
418 ms ± 142 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [19]:
import random
import timeit
code='''
def test_func():
    n=1000
    target=10
    random_arr = random.sample(range(1,n*2),n)
    result2 = search_matched_sum_quickSort(random_arr, target, 0, n-1)
    return result2'''

def test_func():
    n=1000
    target=10
    random_arr = random.sample(range(1,n*2),n)
    result2 = search_matched_sum_quickSort(random_arr, target, 0, n-1)
    return result2

print(test_func())
print(timeit.timeit(stmt=code))
%timeit test_func()

('target=10', [(159, 965), (13, 9)])
0.13661600000000362
130 ms ± 17.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### 3. 배열의 개수가 1000, target이 1000일때 시간을 측정

In [22]:
import random
import timeit
code='''
def test_func():
    n=1000
    target=1000
    random_arr = random.sample(range(1,n*2),n)
    result1 = search_matched_sum_bubbleSort(random_arr,target)
    return result1'''

def test_func():
    n=1000
    target=1000
    random_arr = random.sample(range(1,n*2),n)
    result1 = search_matched_sum_bubbleSort(random_arr,target)
    return result1

print(test_func())
print(timeit.timeit(stmt=code))
%timeit test_func()

('target=1000', [(98, 752), (457, 750), (863, 699), (614, 199), (728, 229), (318, 877), (352, 495), (484, 907), (294, 329), (854, 572), (751, 113), (194, 924), (231, 801), (313, 62), (79, 516), (667, 664), (935, 10), (59, 710), (445, 449), (833, 823), (497, 103), (981, 189), (78, 717), (323, 125), (496, 527), (253, 308), (646, 980), (394, 334), (85, 293), (626, 619), (978, 239), (292, 545), (720, 830), (104, 408), (675, 454), (99, 778), (124, 548), (111, 467), (76, 919), (155, 941), (837, 235), (686, 431), (515, 977), (240, 762), (320, 60), (12, 532), (291, 918), (895, 149), (621, 868), (40, 300), (997, 274), (268, 171), (203, 139), (925, 679), (384, 382), (735, 596), (819, 848), (773, 612), (733, 568), (433, 356), (188, 4), (94, 326), (783, 141), (815, 659), (370, 267), (915, 17), (666, 800), (217, 200), (771, 348), (402, 1), (902, 345), (485, 479), (210, 561), (701, 577), (341, 72), (643, 820), (144, 928), (644, 173), (110, 423), (177, 119), (999, 218), (377, 856), (68, 791), (711, 9

In [23]:
import random
import timeit
code='''
def test_func():
    n=1000
    target=1000
    random_arr = random.sample(range(1,n*2),n)
    result2 = search_matched_sum_quickSort(random_arr, target, 0, n-1)
    return result2'''

def test_func():
    n=1000
    target=1000
    random_arr = random.sample(range(1,n*2),n)
    result2 = search_matched_sum_quickSort(random_arr, target, 0, n-1)
    return result2

print(test_func())
print(timeit.timeit(stmt=code))
%timeit test_func()

('target=1000', [(261, 802), (30, 514), (962, 736), (475, 55), (292, 285), (366, 122), (704, 539), (259, 945), (719, 137), (972, 257), (509, 619), (182, 560), (926, 632), (185, 998), (902, 713), (741, 116), (544, 695), (79, 703), (898, 846), (77, 538), (786, 134), (52, 332), (11, 841), (382, 548), (170, 565), (339, 740), (929, 462), (410, 861), (563, 657), (186, 709), (278, 364), (393, 772), (526, 348), (677, 833), (699, 956), (470, 468), (277, 983), (654, 943), (615, 734), (100, 392), (211, 70), (430, 3), (80, 164), (18, 301), (857, 23), (667, 237), (817, 834), (53, 207), (372, 610), (743, 748), (455, 230), (504, 547), (58, 860), (334, 22), (120, 75), (727, 214), (28, 685), (306, 753), (245, 639), (439, 344), (603, 163), (355, 595), (65, 598), (994, 745), (849, 289), (903, 744), (518, 358), (589, 226), (387, 343), (412, 746), (441, 37), (813, 531), (295, 106), (418, 623), (792, 49), (498, 596), (15, 630), (787, 21), (221, 294), (47, 72), (838, 953), (963, 631), (862, 869), (821, 999),

## ㅁ결과확인 및 정리ㅁ 

### 1. 배열의 개수가 10, target이 10일때는 search_matched_sum_bubbleSort는 평균 61.4 µs,  search_matched_sum_quickSort는 평균 55.8 µs로 사실 거의 차이가 나지않지만 quicksort를 이용한 search가 조금 빨랐음을 확인할 수 있습니다. 2. 배열의 개수가 1000, target이 10일때는 search_matched_sum_bubbleSort는 평균 418 ms,  search_matched_sum_quickSort는 평균 130 ms로 배열의 개수가 10일때보다는 확연한 차이가 발생했음을 알 수 있습니다. 3. 배열의 개수가 1000, target이 1000일때는 search_matched_sum_bubbleSort는 319 ms (1 loop each), search_matched_sum_quickSort는 평균 124 ms (10 loops each)로 역시 확연한 차이가 발생하였습니다. 따라서 세웠던 가설도 해당 실험환경내에서는 맞았음을 확인할 수 있습니다.  